### ISSUE TITLE
Try to use Tree0 for sharing instead of the big [DeltaBind]

### ISSUE THREAD

**Issue by Mikolaj** [Mikolaj/horde-ad#61]
Try to use @tomsmeding's Tree0 (or stable names? we don't have a ticket for that yet, do we? or is this orthogonal?) for sharing instead of the big [DeltaBind].

**Comment by tomsmeding** [comment]
I think stable names are a way to get access to a built-in proxy for heap pointers from GHC. The IDs in Tree and Tree0 are to not have to rely on stable names, and to automatically get a topological order -- with stable names you'd actually have to do a topsort (on the fly in the reverse pass).

Kmett's [ad](https://hackage.haskell.org/package/ad) relies on stable names for this purpose; [Accelerate](https://hackage.haskell.org/package/accelerate) uses stable names too, but for a completely different purpose (to reify user-written let-bindings in the embedded DSL). My experience with them in Accelerate is that they spuriously fail when you use them in parallel enough times for long enough. In Accelerate this manifests as sharing recovery errors if you run multiple instances of the Accelerate compiler in parallel in the same process (uncommon -- the point of Accelerate is to compile _to _ a parallel program -- but it happens in Accelerate's own test suite). It may well be that all those stable names failures are of the sound variety (i.e. the same object yields different names, but "same name => same object" still holds); Accelerate already chokes on those, but for AD it might just be a performance thing.

**Comment by simonpj** [comment]
I would advocate *not* using stable names in the first instance.  Instead:

- The monad allocates fresh DeletaIds (just an Int), but does NOT have a [DeltaBind]
- Instead every Delta node has its DeltaId inside it:
```
data Delta = Add DeltaId Delta Delta 
          | Scale DeltaId Float Delta
          | ....
```
Or, if you prefer
```
data Delta = Delta DeltaId Delta'
data Delta' = Add Delta Delta
             | Scale Float Delta
             | ....
````
I suggest this because it is smaller change; and because stable names carry a pretty heavy overhead.

**Comment by Mikolaj** [comment]
What to do with delta-variables, though, e.g., `Var0` in

https://github.com/Mikolaj/horde-ad/blob/24b591054fdc2bab51f38bc0213b8ed41bc656e7/src/HordeAd/Internal/Delta.hs#L95-L99

Remember that they are dual-purpose: for sharing and in place of of `OneHot`. Do we keep `[DeltaBind]`, but populate it only by non-sharing delta-variables?

**Comment by tomjaguarpaw** [comment]
My original suggestion of `StableName`s was to avoid having to write the surface syntax in a monadic style. If we're happy to continue the latter then no need for `StableNames` whose only benefit would be to recover sharing.

**Comment by Mikolaj** [comment]
I wonder what the long game is. Perhaps, once we reduce the monad via `Tree0` to a fresh name generator, we could replace it with impurely generating the new names and hiding the impurity in a few relevant combinators? This prevents parallel usage of the combinators (similarly as with the monad), but that happens only in the forward pass, which is cheap compared to the backward pass. Not sure how fragile that is once GHC gets creative optimizing.

**Comment by simonpj** [comment]
> My original suggestion of StableNames was to avoid having to write the surface syntax in a monadic style. If we're happy to continue the latter then no need for StableNames whose only benefit would be to recover sharing.

Yes, exactly.  Thanks.

> I wonder what the long game is.

My current thought:
* Experiment with Tree0, as above (not using StableNames). Keep the monadic translation
* Experiment with replacing the monadic translation with StableNames

**Comment by simonpj** [comment]
> What to do with delta-variables, though?

I suggest

```
data Delta0 = Delta0 DeltaId Delta0'

data Delta0 r = 
     Zero0 
   | Scale0 r (Delta0 r) 
   | Add0 (Delta0 r) (Delta0 r) 
   | Input0    -- New
```
where `Input0` says "this is an input".   That should do it.

**Comment by Mikolaj** [comment]
A mock up of an implementation is in #62. Important points:

* to work without a source translation plugin or TH scheme, it requires fresh numbers as impure side effects, which requires disabling some GHC optimizations, disabling concurrent testing and forcing call by value in a lot of code; with plugin or TH we could implement the fresh indices purely by re-introducing a monad
* to handle many ranks efficiently, two counters are required in each `Tree0` node: one exclusive to the rank (to indicate index in the input vectors), the other common for all ranks (to indicate data dependencies and so order of elaboration); I think the same would be needed in the original `Tree`.

BTW, is there any existing library for impure global counters? Ideally with some support from GHC to avoid infecting client code with `-fno-full-laziness -fno-cse`. This is such a common pattern that call-by-value languages provide for free and that could be hidden behind an interface if not for `-fno-full-laziness -fno-cse` (and the existence of global state can't be hidden, but that's the same as using a fixed file in a filesystem, so nothing new in the Haskell world).

**Comment by Mikolaj** [comment]
> The monad allocates fresh DeletaIds (just an Int), but does NOT have a [DeltaBind]

Actually [DeltaBind] resurfaces in the end (or could be built incrementally from the start). Namely, it's needed to keep track of which delta-expression nodes have already been evaluated. Following Tom's paper, I'm implementing it as a mutable array, so it's only a couple times smaller than [DeltaBind] (boxed array vs list). I'd need to benchmark to verify if implementing this as a map and deleting processed entries gives meaningful memory savings in practice and speed from cache locality, etc.

Consequently, the benefit of the Tree0 approach compared to our approach is that it's possible to construct [DeltaBind] only after the forward pass and it occupies less memory. A significant side effect is that the monad (and so the plugin or TH code) can be simpler. The price for that is two counters per node vs one counter in our Var.

**Comment by Mikolaj** [comment]
Actually, the per-rank counters can be assigned after the forward pass. That would require building the [DeltaBind] in one go, in an extra pass over the whole delta expression tree (but not visiting shared nodes more than once, as before). This simplifies the monad to hold only one counter, but the extra traversal is expensive (neither the tree nor the constructed array fit in cache) and requires a few extra screens of code to program the new traversal over the dozens and dozens of constructors we have. Maintaining the extra code is the reason I'd rather not experiment with this variants.

**Comment by Mikolaj** [comment]
I eventually replaced `Input0` with `Zero0`, because while the intention is different, it behaves identically.

The experiment seems relatively complete. I'm going to benchmark now. The only big drawbacks are that it forces any library users to set `-fno-full-laziness -fno-cse` (perhaps it wouldn't if we removed `-fexpose-all-unfoldings` from the topmost modules, but that could slow down user code even more) and also makes it impossible to run many instances of gradient descent in the same process (but it doesn't seem useful except for parallel testing).

**Comment by Mikolaj** [comment]
I've done the benchmarking. Note that this only benchmarks a single data size and also there may be any kind of silly performance mistakes in the code, both branch master and branch 61-Tree0-impure.

The baseline are the numbers from 7 moths ago on the green bars at https://combinatronics.com/Mikolaj/mostly-harmless/bf1014389f69a143bba1525560b9d0b09f0f1a31/misc-bench/MnistAll-efd12c3.html, which are MNIST digit recognition using scalars (rank 0), vectors (rank 1) and matrices (3 matrices and 3 vectors; rank 2 and 1), respectively:

4.37s 245ms 208ms

Current, branch master with GHC 8.10.7 is at (we've definitely broken something around matrices)

4.86s 251ms 436ms

branch master with `-fno-full-laziness -fno-cse` is slower, as expected, but not terribly so

5.45s 249ms 516ms

branch 61-Tree0-impure is at

21.39s 266ms 515ms

Branch 61-Tree0-impure has four times overhead for the neural network defined exclusively with scalar operations on dual numbers. All neural networks in this benchmark have 469160 parameters, which in the other two variants  reside in a single vector or matrix, with a single delta-expression denoting their derivative. But in the variant that uses scalar operations on dual numbers, there is 469160 dual number expressions (and so delta expressions) representing the parameters alone and then as many dual number additions, multiplications, etc.

I suppose the addition of two Int identifiers to each of the delta expressions may increase the size of the delta expressions in memory almost 2 times. Taking cache into account, this may easily be the reason for the 4-fold slowdown. Benchmarking at other sizes and measuring allocation could shed more light on that. It can also be a typo somewhere or trivial memory leak from a thunk. The original code was heavily profiled, so it's easy to break its performance even with small changes and this change was huge.

**Comment by Mikolaj** [comment]
I think I've explained the max residency discrepancy between `master` and `61-Tree0-impure` (the latter is 150% of the former). It does not fully explain 3 times higher allocation and 4-times higher runtime in the particular rank 0 MNIST benchmark I'm now using (this may still be for any silly reasons, though I ruled out obvious memory leaks), but the basic limitations seem clearer now.

The extra max residency comes in half from the per-node identifier, together with related overheads, and in half from the per-rank identifier, which is since today no longer repeated in the nodes, but is still gathered in a mutable array, paired with the relevant subterm (and the pointers to the subterms are duplicated in the array and in the main term). In comparison, on `master` there was perhaps 1 `Var0` subterm per 10000 other subterms, because there were manually simulated `matmult` operations, that is, multiplications and additions of thousands of terms and, obviously, the test code did not assign `Var0` to each `Add0` (though the paper does), etc., but only to the result of the operation. In other words, there was never any "big [DeltaBind]". I have no idea how these would compare if the implementation followed the paper strictly and really created the maximal [DeltaBind].

The fact the problem comes from bulk operations expressed as individual delta terms is why it does not appear in ranks other than rank 0, unless we try to express `map f array` via indexing and it creates thousands of rank 0 terms. But that's a disaster even without each of the rank 0 terms taking 50% more memory.

**Comment by Mikolaj** [comment]
I think PR #62 that implements this idea is ready to merge. It's impure and slow, but it guarantees sharing for any terms re-used by the user without the need for a monad nor any other explicit action by the user (this is what a plugin or TH would give us, but probably with some restrictions and/or unsafety). The impurity is safe, the slowness is only for rank0 (and we are still twice faster than ad IIRC).

While implementing I got the impression the Tree0 formulation is simpler than both a counter *and* the delta bindings in a monad. However, Tree0 gets complicated (and probably less efficient) once we have more than rank 0. The most straightforward port to many ranks would include a second counter, per-rank, in each node. Without it, extra tricks are needed to recover the information. Etc. This is all fresh; I may be missing obvious things.

**Comment by Mikolaj** [comment]
Updated benchmark results are in the #62 PR comments. By hook and by crook I managed to get the runtime of both our benchmarks to be no more than twice worse than master branch. Allocation varies even less and in favour of the PR in some cases.

This Tree0 algorithm has the funny feature that the order of data dependencies encoded in the per-node counters may freely differ from the order in which terms are traversed in the evaluation functions. The thing is, the final phase of evaluation needs to be done strictly in the order of data dependencies. Therefore, if the order of counters and of the traversal differ a lot, computation has to be slow, because a lot has to be stored during traversal, until it can be completely evaluated (but first, looked up in the large data structures). This leads to fragile optimizations that try to align the orders for common cases.
